### DISCUSSION TITLE
How to Validate Checksums Using CRC32C in MinIO with Go for multi-part upload Upload and Download?

### DISCUSSION THREAD

**Discussion by caevv** [minio/minio#18324] [Q&A] [answered]
Hey everyone!

I'm working on a project where I need to ensure data integrity when storing and retrieving large files using MinIO and Go. I've chosen CRC32C for checksum validation. My main question revolves around the PutObject function when it performs multi-part uploads for large files.

Here's a snippet of how I'm currently handling uploads:

```go
func (r *Repository) SaveFile(ctx context.Context, key string, data []byte) error {
	// Calculate CRC32C checksum
	crc := crc32.New(crc32.MakeTable(crc32.Castagnoli))
	if _, err := crc.Write(data); err != nil {
		return err
	}
	meta := map[string]string{"x-amz-checksum-crc32c": base64.StdEncoding.EncodeToString(crc.Sum(nil))}

	// Upload to MinIO
	if _, err := r.client.PutObject(
		ctx,
		"my-bucket",
		key,
		bytes.NewReader(data),
		int64(len(data)),
		minio.PutObjectOptions{
			ContentType:      "application/octet-stream",
			UserMetadata:     meta,
		},
	); err != nil {
		return fmt.Errorf("failed to upload: %w", err)
	}

	return nil
}
```

Or if I change to:
```go
func (r *Repository) SaveFile(ctx context.Context, key string, data io.Reader) error {
	if _, err := r.client.PutObject(
		ctx,
		"my-bucket",
		key,
		data,
		-1,
		minio.PutObjectOptions{},
	); err != nil {
		return fmt.Errorf("failed to upload: %w", err)
	}

	return nil
}
```

Download (is that enough)?
```go
s3Client.GetObject(ctx, bucketName, hashFileBin(blockHash), minio.GetObjectOptions{
	Checksum: true,
})
```

If I `GetObject()` and print `object.Stat()` I get empty checksum in both upload examples:
```
ChecksumCRC32: (string) "",
ChecksumCRC32C: (string) "",
ChecksumSHA1: (string) "",
ChecksumSHA256: (string) "",
``` 

Questions:
- Do I need to implement my checkum as I'm doing on my first example or does it do for me? I don't have a working example of doing a multi-part upload with checksum.
- Multi-part Upload: How does MinIO handle the CRC32C checksum during multi-part uploads? Does it calculate a composite checksum for all parts, or how is this managed?
- Download: What would be the most efficient way to validate the checksum for downloaded files?
- Best Practices: Are there any best practices or common pitfalls I should be aware of when implementing this?

Any guidance would be highly appreciated! 🙏

**Comment by klauspost** [comment, marked as answer]
CRC32C a CRC32C of each part. You cannot use the full file CRC for multipart.

Read https://docs.aws.amazon.com/AmazonS3/latest/userguide/checking-object-integrity.html

`minio-go` will add the CRC32C of each part if no other checksum is sent (`!opts.SendContentMd5 && opts.DisableContentSha256`) if unseekable input is sent or input without size and we have to read parts into memory anyway.

If you set `(minio.Options).TrailingHeaders` on a client it will furthermore use trailing headers for adding CRC32C on streaming uploads (`io.ReaderAt` interface).

We currently don't expose options for checksums in the SDK.

**Comment by caevv** [comment]
If I got your answer right, I don't need to do anything but this:
```go
client.PutObject(
	ctx,
	"my-bucket",
	key,
	data,
	-1,
	minio.PutObjectOptions{
		ContentType:          "application/octet-stream",
		SendContentMd5:       false,
		DisableContentSha256: true,
	},
)
```

```go
client.GetObject(ctx, "my-bucket", key, minio.GetObjectOptions{
	Checksum: true,
})
```

With these options, regardless of being a multi-part upload, it will calculate the checksum for me and data will be integral?

**Reply by klauspost** [reply]
Correct. Multipart checksums will have the number of parts appended. 

This shows how the hash is calculated and combined for multipart:

https://github.com/minio/minio-go/blob/589ff8dc7dd161d4e0ebeb96c2988f055fcec713/functional_tests.go#L2436-L2460
